File: api/Enkl.Api/Services/SamlService.cs
using System.Text.Json;
using Enkl.Api.Auth;
using Enkl.Api.Data;
using Enkl.Api.Domain;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using ITfoxtec.Identity.Saml2;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

public enum SamlAcsOutcome { Success, UserNotFound, UserInactive, JitDisabled }
public record SamlAcsResult(SamlAcsOutcome Outcome, string? ExchangeCode);

/// <summary>
/// SP-side SAML logic shared by SamlController's three actions. Kept separate from the controller
/// because building the AuthnRequest/reading the ACS response is tightly coupled to ITfoxtec's
/// HttpRequest/IActionResult binding helpers (see SamlController), while everything here — org
/// lookup, JIT provisioning, JWT issuance — is plain business logic, same split every other
/// controller/service pair in this codebase already follows.
/// </summary>
public class SamlService
{
    private readonly AppDbContext _db;
    private readonly IConfiguration _config;
    private readonly JwtTokenService _jwt;
    private readonly SsoExchangeCodeStore _exchange;

    public SamlService(AppDbContext db, IConfiguration config, JwtTokenService jwt, SsoExchangeCodeStore exchange)
    {
        _db = db;
        _config = config;
        _jwt = jwt;
        _exchange = exchange;
    }

    private string PublicBaseUrl => (_config["App:PublicBaseUrl"] ?? "").TrimEnd('/');

    public string SpEntityId(Guid orgId) => $"{PublicBaseUrl}/api/saml/{orgId}/metadata";
    public string AcsUrl(Guid orgId) => $"{PublicBaseUrl}/api/saml/{orgId}/acs";
    public string SuccessRedirectUrl(string exchangeCode) => $"{PublicBaseUrl}/?ssoCode={Uri.EscapeDataString(exchangeCode)}";
    public string ErrorRedirectUrl(string message) => $"{PublicBaseUrl}/?ssoError={Uri.EscapeDataString(message)}";

    public async Task<OrganisationSsoConfig?> GetEnabledConfigAsync(Guid orgId)
    {
        var cfg = await _db.OrganisationSsoConfigs.FirstOrDefaultAsync(c => c.OrganisationId == orgId);
        return cfg is { SamlEnabled: true } ? cfg : null;
    }

    public Saml2Configuration BuildSaml2Configuration(Guid orgId, OrganisationSsoConfig ssoConfig)
    {
        var spEntityId = SpEntityId(orgId);
        var config = new Saml2Configuration
        {
            Issuer = spEntityId,
            SingleSignOnDestination = new Uri(ssoConfig.IdpSsoUrl!),
            AllowedIssuer = ssoConfig.IdpEntityId
        };
        config.AllowedAudienceUris.Add(spEntityId);
        if (!string.IsNullOrEmpty(ssoConfig.IdpSigningCertificate) &&
            SamlCertificateHelper.TryParse(ssoConfig.IdpSigningCertificate, out var cert))
        {
            config.SignatureValidationCertificates.Add(cert!);
        }
        return config;
    }

    /// <summary>
    /// Resolves a validated assertion's NameID (email) to a User and, on success, a single-use
    /// exchange code the client trades for a real JWT (see SsoExchangeCodeStore). A signed
    /// assertion for a user of a DIFFERENT organisation is rejected as not-found even though email
    /// is globally unique — defense in depth against a misconfigured/malicious IdP asserting for
    /// this route's org.
    /// </summary>
    public async Task<SamlAcsResult> ProcessAssertionAsync(Guid orgId, OrganisationSsoConfig ssoConfig, string email, string? displayNameHint)
    {
        var normalizedEmail = EmailAddressNormalizer.Normalize(email);
        var user = await _db.Users.Include(u => u.Organisation)
            .FirstOrDefaultAsync(u => u.NormalizedEmailAddress == normalizedEmail);

        if (user is not null && user.OrganisationId != orgId)
        {
            user = null;
        }

        if (user is null)
        {
            if (!ssoConfig.SamlJitProvisioning) return new SamlAcsResult(SamlAcsOutcome.JitDisabled, null);
            user = await JitProvisionUserAsync(orgId, email, normalizedEmail, displayNameHint);
        }

        if (!user.IsActive) return new SamlAcsResult(SamlAcsOutcome.UserInactive, null);

        var memberships = await _db.ProjectMembers.Where(m => m.UserId == user.Id).ToListAsync();
        var (token, expiresAt) = _jwt.GenerateToken(user, memberships);
        var response = new SsoExchangeResponse(token, expiresAt, new UserDto(user.Id, user.Username, user.DisplayName, user.MustChangePassword));
        return new SamlAcsResult(SamlAcsOutcome.Success, _exchange.Issue(JsonSerializer.Serialize(response)));
    }

    private async Task<User> JitProvisionUserAsync(Guid orgId, string email, string normalizedEmail, string? displayNameHint)
    {
        var displayName = string.IsNullOrWhiteSpace(displayNameHint) ? email.Split('@')[0] : displayNameHint.Trim();
        if (displayName.Length > 200) displayName = displayName[..200];

        var baseUsername = UsernameNormalizer.Normalize(displayName.Length > 0 ? displayName : email);
        if (baseUsername.Length == 0) baseUsername = "user";
        var usernameToUse = await _db.Users.AnyAsync(u => u.NormalizedUsername == baseUsername)
            ? await ResolveUniqueUsernameAsync(baseUsername)
            : baseUsername;

        var user = new User
        {
            Id = Guid.NewGuid(),
            OrganisationId = orgId,
            Username = usernameToUse,
            NormalizedUsername = usernameToUse,
            EmailAddress = email,
            NormalizedEmailAddress = normalizedEmail,
            PasswordHash = null,
            DisplayName = displayName,
            MustChangePassword = false,
            IsOrgAdmin = false,
            IsActive = true,
            CreatedAt = DateTime.UtcNow
        };
        _db.Users.Add(user);
        await _db.SaveChangesAsync();
        // JwtTokenService.GenerateToken reads user.Organisation.Name for the orgName claim — the
        // found-user path gets this via .Include() above, so a freshly-created user needs the same
        // navigation loaded explicitly before it's handed to GenerateToken.
        await _db.Entry(user).Reference(u => u.Organisation).LoadAsync();
        return user;
    }

    private async Task<string> ResolveUniqueUsernameAsync(string baseUsername)
    {
        var candidate = baseUsername;
        var suffix = 1;
        while (await _db.Users.AnyAsync(u => u.NormalizedUsername == candidate))
        {
            candidate = $"{baseUsername}{++suffix}";
        }
        return candidate;
    }
}

File: mariadb-api/src/Services/SamlService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Auth\EmailAddressNormalizer;
use Enkl\Api\Auth\JwtService;
use Enkl\Api\Auth\UsernameNormalizer;
use Enkl\Api\Config\Config;
use Enkl\Api\Support\Uuid;
use PDO;

/**
 * Ported from Services/SamlService.cs. SP-side SAML logic shared by SamlController's three
 * actions — org lookup, JIT provisioning, JWT issuance. Kept separate from the controller because
 * SamlController's actions are tightly coupled to onelogin/php-saml's Auth class and PHP
 * superglobals (see that file's own comment on why the ACS action reads $_POST directly), while
 * everything here is plain business logic, same split every other controller/service pair in this
 * codebase already follows.
 */
final class SamlService
{
    public function __construct(
        private readonly PDO $db,
        private readonly SsoExchangeCodeService $exchange
    ) {
    }

    private function publicBaseUrl(): string
    {
        return rtrim((string) Config::get('APP_PUBLIC_BASE_URL', ''), '/');
    }

    public function spEntityId(string $orgId): string
    {
        return $this->publicBaseUrl() . '/api/saml/' . $orgId . '/metadata';
    }

    public function acsUrl(string $orgId): string
    {
        return $this->publicBaseUrl() . '/api/saml/' . $orgId . '/acs';
    }

    public function successRedirectUrl(string $exchangeCode): string
    {
        return $this->publicBaseUrl() . '/?ssoCode=' . rawurlencode($exchangeCode);
    }

    public function errorRedirectUrl(string $message): string
    {
        return $this->publicBaseUrl() . '/?ssoError=' . rawurlencode($message);
    }

    /** @return array<string,mixed>|null */
    public function getEnabledConfig(string $orgId): ?array
    {
        $stmt = $this->db->prepare('SELECT * FROM "OrganisationSsoConfigs" WHERE "OrganisationId" = :id');
        $stmt->execute(['id' => $orgId]);
        $cfg = $stmt->fetch();
        if ($cfg === false || !$cfg['SamlEnabled']) {
            return null;
        }
        return $cfg;
    }

    /**
     * Builds the onelogin/php-saml settings array for actual IdP-facing operations (login/acs) —
     * the metadata action needs no IdP fields at all, see SamlController::metadata's own settings.
     * @param array<string,mixed> $ssoConfig
     * @return array<string,mixed>
     */
    public function buildAuthSettings(string $orgId, array $ssoConfig): array
    {
        return [
            'strict' => true,
            'sp' => [
                'entityId' => $this->spEntityId($orgId),
                'assertionConsumerService' => [
                    'url' => $this->acsUrl($orgId),
                    'binding' => 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST',
                ],
                'NameIDFormat' => 'urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress',
            ],
            'idp' => [
                'entityId' => $ssoConfig['IdpEntityId'],
                'singleSignOnService' => [
                    'url' => $ssoConfig['IdpSsoUrl'],
                    'binding' => 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect',
                ],
                // Utils::formatCert() (called internally by the library) accepts either a bare
                // base64 DER string or an already-PEM-wrapped certificate — the raw stored value is
                // passed through unchanged; see SamlCertificateHelper's own comment.
                'x509cert' => $ssoConfig['IdpSigningCertificate'] ?? '',
            ],
            'security' => [
                'authnRequestsSigned' => false,
                'wantAssertionsSigned' => true,
                'wantNameId' => true,
            ],
        ];
    }

    /**
     * Resolves a validated assertion's NameID (email) to a User and, on success, a single-use
     * exchange code the client trades for a real JWT (see SsoExchangeCodeService). A signed
     * assertion for a user of a DIFFERENT organisation is rejected as not-found even though email
     * is globally unique — defense in depth against a misconfigured/malicious IdP asserting for
     * this route's org.
     * @param array<string,mixed> $ssoConfig
     * @return array{outcome:string, exchangeCode:?string}
     */
    public function processAssertion(string $orgId, array $ssoConfig, string $email, ?string $displayNameHint): array
    {
        $normalizedEmail = EmailAddressNormalizer::normalize($email);
        $stmt = $this->db->prepare(<<<SQL
            SELECT u.*, o."Name" AS "OrganisationName", p."Avatar", p."HeaderColour" FROM "Users" u
            JOIN "Organisations" o ON o."Id" = u."OrganisationId"
            LEFT JOIN "UserPreferences" p ON p."UserId" = u."Id"
            WHERE u."NormalizedEmailAddress" = :n LIMIT 1
        SQL);
        $stmt->execute(['n' => $normalizedEmail]);
        $user = $stmt->fetch();

        if ($user !== false && $user['OrganisationId'] !== $orgId) {
            $user = false;
        }

        if ($user === false) {
            if (!$ssoConfig['SamlJitProvisioning']) {
                return ['outcome' => 'jit_disabled', 'exchangeCode' => null];
            }
            $user = $this->jitProvisionUser($orgId, $email, $normalizedEmail, $displayNameHint);
        }

        if (!$user['IsActive']) {
            return ['outcome' => 'user_inactive', 'exchangeCode' => null];
        }

        $stmt = $this->db->prepare('SELECT "ProjectId", "Role", "IsProjectAdmin" FROM "ProjectMembers" WHERE "UserId" = :uid');
        $stmt->execute(['uid' => $user['Id']]);
        $memberships = $stmt->fetchAll();

        $tokenInfo = JwtService::generateToken($user, $memberships);
        $payload = json_encode([
            'token' => $tokenInfo['token'],
            'expiresAt' => $tokenInfo['expiresAt'],
            'user' => [
                'id' => $user['Id'],
                'username' => $user['Username'],
                'displayName' => $user['DisplayName'],
                // MariaDB port: PDO_MYSQL returns a BOOLEAN/TINYINT(1) column as a plain PHP int
                // (0/1), never a native bool the way PDO_PGSQL always does — explicit cast keeps
                // this real JSON true/false, matching the other two tiers' response shape.
                'mustChangePassword' => (bool) $user['MustChangePassword'],
                'avatar' => $user['Avatar'] ?? null,
                'headerColour' => $user['HeaderColour'] ?? null,
            ],
        ]);

        return ['outcome' => 'success', 'exchangeCode' => $this->exchange->issue((string) $payload)];
    }

    /** @return array<string,mixed> the created user row (including OrganisationName, for JwtService::generateToken) */
    private function jitProvisionUser(string $orgId, string $email, string $normalizedEmail, ?string $displayNameHint): array
    {
        $displayName = trim((string) $displayNameHint) !== '' ? trim((string) $displayNameHint) : explode('@', $email)[0];
        if (strlen($displayName) > 200) {
            $displayName = substr($displayName, 0, 200);
        }

        $baseUsername = UsernameNormalizer::normalize($displayName !== '' ? $displayName : $email);
        if ($baseUsername === '') {
            $baseUsername = 'user';
        }
        $usernameToUse = $this->resolveUniqueUsername($baseUsername);

        $userId = Uuid::v4();
        // MariaDB port: "SecurityStamp" has no DB-side default here (see
        // src/Db/migrations/001_initial_schema.sql's own note) — unlike php-api's original, which
        // relied on Postgres's `DEFAULT gen_random_uuid()`, this INSERT must supply one explicitly.
        $stmt = $this->db->prepare(<<<SQL
            INSERT INTO "Users" ("Id", "OrganisationId", "Username", "NormalizedUsername", "EmailAddress", "NormalizedEmailAddress", "PasswordHash", "DisplayName", "MustChangePassword", "IsOrgAdmin", "IsActive", "CreatedAt", "SecurityStamp")
            VALUES (:id, :orgId, :username, :normalized, :email, :normalizedEmail, NULL, :displayName, false, false, true, now(), :securityStamp)
        SQL);
        $stmt->execute([
            'id' => $userId, 'orgId' => $orgId, 'username' => $usernameToUse, 'normalized' => $usernameToUse,
            'email' => $email, 'normalizedEmail' => $normalizedEmail, 'displayName' => $displayName,
            'securityStamp' => Uuid::v4(),
        ]);

        $stmt = $this->db->prepare(<<<SQL
            SELECT u.*, o."Name" AS "OrganisationName" FROM "Users" u
            JOIN "Organisations" o ON o."Id" = u."OrganisationId"
            WHERE u."Id" = :id
        SQL);
        $stmt->execute(['id' => $userId]);
        return $stmt->fetch();
    }

    private function resolveUniqueUsername(string $baseUsername): string
    {
        $candidate = $baseUsername;
        $suffix = 1;
        $stmt = $this->db->prepare('SELECT 1 FROM "Users" WHERE "NormalizedUsername" = :n');
        do {
            $stmt->execute(['n' => $candidate]);
            if ($stmt->fetch() === false) {
                break;
            }
            $suffix++;
            $candidate = $baseUsername . $suffix;
        } while (true);
        return $candidate;
    }
}
